### PR TITLE
Add TaylorSeriesRationalFunction and improve Derivative for univariate rational functions

### PR DESCRIPTION
**Pull request by hulpke**
Includes better method for derivative of rational function.

Taylor series coefficients may be useful for Molien series or Growth series

### CODE REVIEW

**Comment by fingolfin** [comment]
This is a new documented feature, so it does need release notes. I've changed the PR title to a suitable candidate for the release notes entry.

File: lib/ratfunul.gi [hunk 1]
```diff
@@ -1069,18 +1069,72 @@ end);
...
   [IsPolynomial],[IsUnivariateRationalFunction],0);
 
+
+InstallGlobalFunction(TaylorSeriesRationalFunction,function(f,at,deg)
+local t,i,x;
+  if not IsUnivariateRationalFunction(f) then
+    Error("function is not univariate");
+  fi;
+  x:=X(LeftActingDomain(DefaultRing(f^0)),
+    IndeterminateNumberOfUnivariateRationalFunction(f));
```
Comment (fingolfin) [inline lib/ratfunul.gi:1127]:
```suggestion
  x:=IndeterminateOfUnivariateRationalFunction(f);
```

File: lib/ratfun.gd [hunk 2]
```diff
@@ -1340,6 +1340,27 @@ DeclareOperation("DegreeIndeterminate",[IsPolynomial,IsPosInt]);
...
+##
+##  <#GAPDoc Label="TaylorSeriesRationalFunction">
+##  <ManSection>
+##  <Attr Name="TaylorSeriesRationalFunction" Arg='ratfun, at, deg]'/>
+##
+##  <Description>
+##  Computes the taylor series up to degree <A>deg</A> of <A>ratfun</A> at
+##  <A>at</A>.
+##  <Example><![CDATA[
+##  gap> TaylorSeriesRationalFunction( (x^5+3*x+7)/(x^5+x+1) ,0,11);
```
Comment (fingolfin) [inline lib/ratfun.gd:1355]:
```suggestion
##  gap> TaylorSeriesRationalFunction((x^5+3*x+7)/(x^5+x+1),0,11);
```

File: lib/ratfunul.gi [hunk 3]
```diff
@@ -1069,18 +1069,72 @@ end);
...
 
 RedispatchOnCondition(Derivative,true,
   [IsPolynomial],[IsUnivariateRationalFunction],0);
 
+
+InstallGlobalFunction(TaylorSeriesRationalFunction,function(f,at,deg)
+local t,i,x;
+  if not IsUnivariateRationalFunction(f) then
+    Error("function is not univariate");
+  fi;
```
Comment (fingolfin) [inline lib/ratfunul.gi:1125]: Perhaps also check that `deg` is a non-negative integer?

File: lib/ratfunul.gi [hunk 4]
```diff
@@ -1069,18 +1069,72 @@ end);
...
+  x:=X(LeftActingDomain(DefaultRing(f^0)),
+    IndeterminateNumberOfUnivariateRationalFunction(f));
+  t:=Zero(f);
+  for i in [0..deg] do
+    if i>0 then 
+      f:=Derivative(f);
+    fi;
+    t:=t+Value(f,at)*(x-at)^i/Factorial(i);
+  od;
```
Comment (fingolfin) [inline lib/ratfunul.gi:1134]: Assuming we require `deg >= 0` (which is not clear to me from the description of the function), could also do this
```suggestion
  t:=Value(f,at)*x^0;
  for i in [1..deg] do
    f:=Derivative(f);
    t:=t+Value(f,at)*(x-at)^i/Factorial(i);
  od;
```

**Review by fingolfin — APPROVED** [review]
Looks OK in principle but I have some suggestions. But it's also OK to merge as-is *shrug*

File: lib/ratfunul.gi (hunk 4 above)
Comment (hulpke) [reply lib/ratfunul.gi:1123]: I think this is a wash in terms of cost or code readability, so I left it out. (Added test for deg\ge0)

**Pull request merged**